*: add foreign key doc (#11974)
pingcap/docs-cn: sql-statements/sql-statement-foreign-key.md

---
title: FOREIGN KEY
summary: TiDB 数据库中 FOREIGN KEY 的使用概况。
---

# FOREIGN KEY

> 外键是 TiDB 从 v6.5.0 开始支持的**实验特性**，允许跨表交叉引用相关数据，以及外键约束，有助于保持相关数据的一致性。

Reviewer: 应该是 6.6 吧，同时也不用实验特性了，DMR 可以直接带出去好了，默认打开。

外键是在子表中定义的，语法如下：

```sql
[CONSTRAINT identifier] FOREIGN KEY
    [identifier] (col_name, ...)
    REFERENCES tbl_name (col_name,...)
    [ON DELETE reference_option]
    [ON UPDATE reference_option]

reference_option:
    RESTRICT | CASCADE | SET NULL | NO ACTION | SET DEFAULT
```

> **警告：**
> 
> 由于外键还是实验特性，所有并没有默认开启该功能，需要手动设置以下变量后打开该功能：

> > `set @@global.tidb_enable_foreign_key=1;`

Reviewer: 这个参数可以删除了，只需要 foreign_key_checks 来控制约束校验是否生效即可

---
title: FOREIGN KEY
summary: TiDB 数据库中 FOREIGN KEY 的使用概况。
---

# FOREIGN KEY

外键是 TiDB 从 v6.6.0 开始支持的功能，允许跨表交叉引用相关数据，以及外键约束，有助于保持相关数据的一致性。

> 外键是在子表中定义的，语法如下：

Reviewer 2: ”子表“ 这个概念不是特别合适？记得 pg 有专门的 inherited table 一章（里面有述 parent table 和 child table），mysql 应该没有这个概念，最好这里用约束表和被约束表/引用表和被引用表？或者统一在这边 declare 一下：父子表的指代。

Author: MySQL 的文档里面也是用的 child table 来描述的：https://dev.mysql.com/doc/refman/8.0/en/create-table-foreign-keys.html

---

> title: FOREIGN KEY

Reviewer 3: Please add this newly added document to `TOC.md`.

Author: great, done.

    [ON DELETE reference_option]
    [ON UPDATE reference_option]
```

外键可以是自引用的，即引用同一个表。使用 `ALTER TABLE` 向表添加外键约束时，请先在外键引用父表的列上创建索引。

## 删除外键约束

可以使用下面 `ALTER TABLE` 语句来删除一个外键约束：

```sql
ALTER TABLE table_name DROP FOREIGN KEY fk_identifier;
```

如果外键约束在创建时定义了名称，则可以引用该名称来删除外键约束。否则，约束名称是自动生成的。你可以使用 `SHOW CREATE TABLE` 查看外键名称：

```sql
mysql> SHOW CREATE TABLE child\G
*************************** 1. row ***************************
       Table: child
Create Table: CREATE TABLE `child` (
  `id` int(11) DEFAULT NULL,
  `pid` int(11) DEFAULT NULL,
  KEY `idx_pid` (`pid`),
  CONSTRAINT `fk_1` FOREIGN KEY (`pid`) REFERENCES `test`.`parent` (`id`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```

## 外键约束检查

TiDB 支持是否开启外键约束检查，由系统变量 [`foreign_key_checks`](/system-variables.md#foreign_key_checks) 控制，其默认值是 `ON`，即开启外键约束检查，它有 `GLOBAL` 和 `SESSION` 两种作用域。在一般的操作中保持该变量开启可以保证外键引用关系的完整性。

关闭 [`foreign_key_checks`](/system-variables.md#foreign_key_checks) 的作用如下：

- 当删除一个被外键引用的父表时，只有关闭外键约束检查时才能删除成功。
- 当给数据库导入数据时，创建表的顺序可能和外键依赖顺序不一样而导致创建表报错，只有关闭外键约束检查时才能创建表成功，另外，导入数据时关闭外键约束检查也能加快导数据的速度。
- 当给数据库导入数据时，先导入子表的数据会报错，只有关闭外键约束检查，才能确保顺利导入子表数据。

> - 执行有关外键的 `ALTER TABLE` 操作时，关于外键约束检查才能执行成功。

Author: ```suggestion
- 执行有关外键的 `ALTER TABLE` 操作时，关闭外键约束检查才能执行成功。
```

> - 执行有关外键的 `ALTER TABLE` 操作时，关于外键约束检查才能执行成功。

Reviewer 4: ”关于“改成”关闭“？

Author: nice catch, done.

当关闭关键约束检查时，不会执行外键约束检查以及引用操作，但以下场景除外：

> - `ALTER TABLE` 执行时如果导致外键的定义不正确，则依然会执行报错。

Reviewer 4: ```suggestion
- 如果执行 `ALTER TABLE` 会导致外键定义不正确，则依然会执行报错。
```

- 删除外键所需的索引时，需要先删除外键，否则删除外键会执行报错。
- 创建外键时，如果不符合外键的条件或限制，则依然会执行报错。

## 锁

> 在 `INSERT` 或者 `UPDATE` 子表时，外键约束会检查父表中是否存在对应的外键值，并对父表中的该行数据上锁，避免该外键值被其他操作删除，导致破坏外键约束。这里的上锁行为等同于对父表中外键值所在行做 `SELECT FOR UPDATE` 操作。因为 TiDB 目前暂不支持 `LOCK IN SHARE MODE`，所以如果在并发写入子表且引用的外键值大部分都一样时，可能会有比较严重的锁冲突，建议在大批量写入子表数据时，关闭 [`foreign_key_checks`](/system-variables.md#foreign_key_checks)。

Reviewer 4: ```suggestion
在 `INSERT` 或者 `UPDATE` 子表时，外键约束会检查父表中是否存在对应的外键值，并对父表中的该行数据上锁，避免该外键值被其他操作删除，导致破坏外键约束。这里的上锁行为等同于对父表中外键值所在行做 `SELECT FOR UPDATE` 操作。因为 TiDB 目前暂不支持 `LOCK IN SHARE MODE`，所以，在并发写入子表场景，如果引用的外键值大部分都一样，可能会有比较严重的锁冲突。建议在大批量写入子表数据时，关闭 [`foreign_key_checks`](/system-variables.md#foreign_key_checks)。
```

- 如果 `CONSTRAINT identifier` 和 `FOREIGN KEY identifier` 语句都没有指定名称，则会自动生成一个名称，例如 `fk_1`、`fk_2`、`fk_3` 等。
- 外键名称必须在当前表中唯一，否则创建时会报错 `ERROR 1826: Duplicate foreign key constraint name 'fk'`。

## 限制

创建外键时需要满足以下条件：

- 父表和子表都不能是临时表。
- 用户需要对父表有 `REFERENCES` 权限。
- 父表和子表中外键引用的列必须是相同的数据类型，并具有相同的大小、精度、长度、字符集 (charset) 和排序规则 (collation)。
- 外键中的列不能引用自身。
- 外键中的列和引用的父表中的列必须有对应的索引，并且索引中的列顺序必须与外键的列顺序一样，这样才能在执行外键约束检查时使用索引来避免全表扫描。

    - 如果父表中没有对应的外键索引，则会报错 `ERROR 1822: Failed to add the foreign key constraint. Missing index for constraint 'fk' in the referenced table 't'`。
    - 如果子表中没有对应的外键索引，则会自动创建一个索引，索引名和外键名一样。

- 不支持在 `BLOB` 和 `TEXT` 类型的列上创建外键。
- 不支持在分区表上创建外键。
- 不支持在虚拟生成列 (`VIRTUAL GENERATED COLUMNS`) 上创建外键。

## 引用操作

当 `UPDATE` 或 `DELETE` 操作影响父表中的外键值时，其在子表中相匹配的外键值取决于外键定义中 `ON UPDATE` 和 `ON DELETE` 定义的引用操作，引用操作包括：

- `CASCADE`：当 `UPDATE` 或 `DELETE` 父表中的行数据时，自动级联更新或删除子表中的匹配行数据。级联操作会用深度优先方式执行。
- `SET NULL`：当 `UPDATE` 或 `DELETE` 父表中的行数据时，自动将子表中匹配的外键列数据设置为 `NULL`。
- `RESTRICT`：如果子表中存在外键匹配的行数据，则拒绝 `UPDATE` 或 `DELETE` 父表的操作。
- `NO ACTION`：行为和 `RESTRICT` 一样。
- `SET DEFAULT`：行为和 `RESTRICT` 一样。

如果父表中没有匹配的外键值，则拒绝 `INSERT` 或 `UPDATE` 子表的操作。

如果外键定义中没有指定 `ON DELETE` 或者 `ON UPDATE`，则默认的行为是 `NO ACTION`。

> 如果外键是定义在 `STORED GENERATED COLUMN` 上的，则不支持使用 `CASCADE`，`SET NULL` 和 `SET DEFAULT` 引用操作。

Reviewer 3: ```suggestion
如果外键是定义在 `STORED GENERATED COLUMN` 上的，则不支持使用 `CASCADE`、`SET NULL` 和 `SET DEFAULT` 引用操作。
```

## 外键使用示例

下面的示例通过单列外键关联父表和子表：

> 
> {{< copyable "sql" >}}
> 

Reviewer 3: ```suggestion

```

```sql
CREATE TABLE parent (
    id INT KEY
);

CREATE TABLE child (
    id INT,
    pid INT,
    INDEX idx_pid (pid),
    FOREIGN KEY (pid) REFERENCES parent(id) ON DELETE CASCADE
);
```

> 下面是一个更复杂的示例，其中 `product_order` 表有 2 个外键分别引用其他两个表。一个外键引用 `product` 表中的两列索引。另一个引用 `customer` 表中的单列索引：

Reviewer 3: ```suggestion
下面是一个更复杂的示例，其中 `product_order` 表有两个外键分别引用其他两个表。一个外键引用 `product` 表中的两列索引。另一个引用 `customer` 表中的单列索引：
```

> 
> {{< copyable "sql" >}}
> 

Reviewer 3: ```suggestion

```

```sql
CREATE TABLE product (
    category INT NOT NULL, 
    id INT NOT NULL,
    price DECIMAL(20,10),
    PRIMARY KEY(category, id)
);

CREATE TABLE customer (
    id INT KEY
);

CREATE TABLE product_order (
    id INT NOT NULL AUTO_INCREMENT,
    product_category INT NOT NULL,
    product_id INT NOT NULL,
    customer_id INT NOT NULL,

    PRIMARY KEY(id),
    INDEX (product_category, product_id),
    INDEX (customer_id),

    FOREIGN KEY (product_category, product_id)
      REFERENCES product(category, id)
      ON UPDATE CASCADE ON DELETE RESTRICT,

    FOREIGN KEY (customer_id)
      REFERENCES customer(id)
);
```

## 新增外键约束

可以使用下面 `ALTER TABLE` 语句来新增一个外键约束：

```sql
ALTER TABLE table_name
    ADD [CONSTRAINT [identifier]] FOREIGN KEY
    [identifier] (col_name, ...)
    REFERENCES tbl_name (col_name,...)
    [ON DELETE reference_option]
    [ON UPDATE reference_option]
```

> 外键可以是自引用的（引用同一个表）。使用 ALTER TABLE 向表添加外键约束时，请先在外键引用父表的列上创建索引。

Reviewer 3: ```suggestion
外键可以是自引用的，即引用同一个表。使用 `ALTER TABLE` 向表添加外键约束时，请先在外键引用父表的列上创建索引。
```

## 删除外键约束

可以使用下面 `ALTER TABLE` 语句来删除一个外键约束：

```sql
ALTER TABLE table_name DROP FOREIGN KEY fk_identifier;
```

> 如果外键约束在创建时定义了名称，则可以引用该名称来删除外键约束。否则约束名称是自动生成的，您可以使用 `SHOW CREATE TABLE` 查看外键名称：

Reviewer 3: ```suggestion
如果外键约束在创建时定义了名称，则可以引用该名称来删除外键约束。否则，约束名称是自动生成的。你可以使用 `SHOW CREATE TABLE` 查看外键名称：
```

  `id` int(11) DEFAULT NULL,
  `pid` int(11) DEFAULT NULL,
  KEY `idx_pid` (`pid`),
  CONSTRAINT `fk_1` FOREIGN KEY (`pid`) REFERENCES `test`.`parent` (`id`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```

## 外键约束检查

TiDB 支持是否开启外键约束检查，由系统变量 [`foreign_key_checks`](/system-variables.md#foreign_key_checks) 控制，其默认值是 `ON`，即开启外键约束检查，它有 `GLOBAL` 和 `SESSION` 两种作用域。在一般的操作中保持该变量开启可以保证外键引用关系的完整性。

关闭 [`foreign_key_checks`](/system-variables.md#foreign_key_checks) 的作用如下：

- 当删除一个被外键引用的父表时，只有关闭外键约束检查时才能删除成功。
- 当给数据库导入数据时，创建表的顺序可能和外键依赖顺序不一样而导致创建表报错，只有关闭外键约束检查时才能创建表成功，另外，导入数据时关闭外键约束检查也能加快导数据的速度。
- 当给数据库导入数据时，如果先导入子表的数据会报错，只有关闭外键约束检查时才能成功先导入子表的数据。
- 执行有关外键的 `ALTER TABLE` 操作时，关于外键约束检查才能执行成功。

当关闭关键约束检查时，不会执行外键约束检查以及引用操作，但以下场景除外：

- `ALTER TABLE` 执行时如果导致外键的定义不正确，则依然会执行报错。
- 删除外键所需的索引时，需要先删除外键，否则删除外键会执行报错。
- 创建外键时，如果不符合外键的条件或限制，则依然会执行报错。

## 锁

在 `INSERT` 或者 `UPDATE` 子表时，外键约束会检查父表中是否存在对应的外键值，并对父表中的该行数据上锁，避免该外键值被其他操作删除，导致破坏外键约束。这里的上锁行为等同于对父表中外键值所在行做 `SELECT FOR UPDATE` 操作。因为 TiDB 目前暂不支持 `LOCK IN SHARE MODE`，所以如果在并发写入子表且引用的外键值大部分都一样时，可能会有比较严重的锁冲突，建议在大批量写入子表数据时，关闭 [`foreign_key_checks`](/system-variables.md#foreign_key_checks)。

## 外键的定义和元信息

> 您可以使用 `SHOW CREATE TABLE` 语句查看外键的定义：

Reviewer 3: ```suggestion
你可以使用 `SHOW CREATE TABLE` 语句查看外键的定义：
```

```sql
mysql> SHOW CREATE TABLE child\G
*************************** 1. row ***************************
       Table: child
Create Table: CREATE TABLE `child` (
  `id` int(11) DEFAULT NULL,
  `pid` int(11) DEFAULT NULL,
  KEY `idx_pid` (`pid`),
  CONSTRAINT `fk_1` FOREIGN KEY (`pid`) REFERENCES `test`.`parent` (`id`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```

> 您可以从 `INFORMATION_SCHEMA.KEY_COLUMN_USAGE` 系统表中获取有关的外键信息，下面是一个查询示例：

Reviewer 3: ```suggestion
你可以从 `INFORMATION_SCHEMA.KEY_COLUMN_USAGE` 系统表中获取有关的外键信息，下面是一个查询示例：
```

```sql
mysql> SELECT TABLE_SCHEMA, TABLE_NAME, COLUMN_NAME, CONSTRAINT_NAME FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE WHERE REFERENCED_TABLE_SCHEMA IS NOT NULL;
+--------------+---------------+------------------+-----------------+
| TABLE_SCHEMA | TABLE_NAME    | COLUMN_NAME      | CONSTRAINT_NAME |
+--------------+---------------+------------------+-----------------+
| test         | child         | pid              | fk_1            |
| test         | product_order | product_category | fk_1            |
| test         | product_order | product_id       | fk_1            |
| test         | product_order | customer_id      | fk_2            |
+--------------+---------------+------------------+-----------------+
```

## 查看带有外键的执行计划

> 您可以使用 `EXPLAIN` 语句查看执行计划。`Foreign_Key_Check` 算子是执行 DML 语句时，执行外键约束检查的算子。

Reviewer 3: ```suggestion
你可以使用 `EXPLAIN` 语句查看执行计划。`Foreign_Key_Check` 算子是执行 DML 语句时，执行外键约束检查的算子。
```

```sql
mysql> explain insert into child values (1,1);
+-----------------------+---------+------+---------------+-------------------------------+
| id                    | estRows | task | access object | operator info                 |
+-----------------------+---------+------+---------------+-------------------------------+
| Insert_1              | N/A     | root |               | N/A                           |
| └─Foreign_Key_Check_3 | 0.00    | root | table:parent  | foreign_key:fk_1, check_exist |
+-----------------------+---------+------+---------------+-------------------------------+
```

> 您可以使用 `EXPLAIN ANALYZE` 语句查看外键引用行为的执行。`Foreign_Key_Cascade` 算子是执行 DML 语句时，执行外键引用行为的算子。

Reviewer 3: ```suggestion
你可以使用 `EXPLAIN ANALYZE` 语句查看外键引用行为的执行。`Foreign_Key_Cascade` 算子是执行 DML 语句时，执行外键引用行为的算子。
```

|       ├─IndexRangeScan_9(Build)  | 10.00   | 1       | cop[tikv] | table:child, index:idx_pid(pid) | time:351.2µs, loops:3, cop_task: {num: 1, max: 282.3µs, proc_keys: 0, rpc_num: 1, rpc_time: 263µs, copr_cache_hit_ratio: 0.00, distsql_concurrency: 15}, tikv_task:{time:220.2µs, loops:0}   | range:[1,1], keep order:false, stats:pseudo | N/A       | N/A  |
|       └─TableRowIDScan_10(Probe) | 10.00   | 1       | cop[tikv] | table:child                     | time:223.9µs, loops:2, cop_task: {num: 1, max: 168.8µs, proc_keys: 0, rpc_num: 1, rpc_time: 154.5µs, copr_cache_hit_ratio: 0.00, distsql_concurrency: 15}, tikv_task:{time:145.6µs, loops:0} | keep order:false, stats:pseudo              | N/A       | N/A  |
+----------------------------------+---------+---------+-----------+---------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------+-----------+------+
```

## 兼容性

> ### 与旧版本的 TiDB 兼容性

Reviewer 3: ```suggestion
### TiDB 版本间兼容性
```

> 由于 TiDB 在 v6.6.0 版本之前已经支持创建外键的语法，但创建的外键并不生效的。在用户将 TiDB 升级到 v6.6.0 版本及以后，之前在旧版本上创建的外键依然是不生效的，只有在v6.6.0 版本及以后新创建的外键才生效。你可以使用 `SHOW CREATE TABLE` 语句查看外键是否生效，不生效的外键会有一条 `/* FOREIGN KEY INVALID */` 注释。

Reviewer 3: ```suggestion
TiDB 在 v6.6.0 之前已经支持创建外键的语法，但创建的外键并不生效。如果将之前创建的 TiDB 集群升级到 v6.6.0 及之后的版本，之前创建的外键依然是不生效的。只有在 v6.6.0 及之后版本中新创建的外键才生效。你可以使用 `SHOW CREATE TABLE` 语句查看外键是否生效，不生效的外键会有一条 `/* FOREIGN KEY INVALID */` 注释。
```

```sql
ALTER TABLE table_name DROP FOREIGN KEY fk_identifier;
```

如果外键约束在创建时定义了名称，则可以引用该名称来删除外键约束。否则，约束名称是自动生成的。你可以使用 `SHOW CREATE TABLE` 查看外键名称：

```sql
mysql> SHOW CREATE TABLE child\G
*************************** 1. row ***************************
       Table: child
Create Table: CREATE TABLE `child` (
  `id` int(11) DEFAULT NULL,
  `pid` int(11) DEFAULT NULL,
  KEY `idx_pid` (`pid`),
  CONSTRAINT `fk_1` FOREIGN KEY (`pid`) REFERENCES `test`.`parent` (`id`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```

## 外键约束检查

TiDB 支持是否开启外键约束检查，由系统变量 [`foreign_key_checks`](/system-variables.md#foreign_key_checks) 控制，其默认值是 `ON`，即开启外键约束检查，它有 `GLOBAL` 和 `SESSION` 两种作用域。在一般的操作中保持该变量开启可以保证外键引用关系的完整性。

关闭 [`foreign_key_checks`](/system-variables.md#foreign_key_checks) 的作用如下：

- 当删除一个被外键引用的父表时，只有关闭外键约束检查时才能删除成功。
- 当给数据库导入数据时，创建表的顺序可能和外键依赖顺序不一样而导致创建表报错，只有关闭外键约束检查时才能创建表成功，另外，导入数据时关闭外键约束检查也能加快导数据的速度。
- 当给数据库导入数据时，先导入子表的数据会报错，只有关闭外键约束检查，才能确保顺利导入子表数据。
- 执行有关外键的 `ALTER TABLE` 操作时，关闭外键约束检查才能执行成功。

当关闭关键约束检查时，不会执行外键约束检查以及引用操作，但以下场景除外：

- 如果执行 `ALTER TABLE` 会导致外键定义不正确，则依然会执行报错。
- 删除外键所需的索引时，需要先删除外键，否则删除外键会执行报错。
- 创建外键时，如果不符合外键的条件或限制，则依然会执行报错。

> ## 锁

Author: 暂时想不到别的名字了，MySQL 也是这个标题的，https://dev.mysql.com/doc/refman/8.0/en/create-table-foreign-keys.html#foreign-key-locking

+----------------------------------+---------+---------+-----------+---------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------+-----------+------+
```

## 兼容性

### TiDB 版本间兼容性

TiDB 在 v6.6.0 之前已经支持创建外键的语法，但创建的外键并不生效。如果将之前创建的 TiDB 集群升级到 v6.6.0 及之后的版本，之前创建的外键依然是不生效的。只有在 v6.6.0 及之后版本中新创建的外键才生效。你可以使用 `SHOW CREATE TABLE` 语句查看外键是否生效，不生效的外键会有一条 `/* FOREIGN KEY INVALID */` 注释。

```sql
mysql> SHOW CREATE TABLE child\G
***************************[ 1. row ]***************************
Table        | child
Create Table | CREATE TABLE `child` (
  `id` int(11) DEFAULT NULL,
  `pid` int(11) DEFAULT NULL,
  KEY `idx_pid` (`pid`),
  CONSTRAINT `fk_1` FOREIGN KEY (`pid`) REFERENCES `test`.`parent` (`id`) ON DELETE CASCADE /* FOREIGN KEY INVALID */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```

> ### 与 TiDB 工具的兼容性

Author: @lance6716 @overvenus @3pointer PTAL

Reviewer 5: BR part LGTM

- [TiDB Binlog](/tidb-binlog/tidb-binlog-overview.md) 不支持外键功能。

> - [DM](/dm/dm-overview.md) 在同步数据到下游 TiDB 时，会始终关闭下游 TiDB 的 `foreign_key_checks`，所以由外键产生级联操作不会从上游同步到下游，进而导致上下游数据不一致。这与旧版本 DM 的行为一致，因为在旧版本中 TiDB 不支持外键。

Reviewer 6: ```suggestion
- v6.6.0 版本的 [DM](/dm/dm-overview.md) 在同步数据到下游 TiDB 时，会显式关闭下游 TiDB 的 `foreign_key_checks`，所以由外键产生级联操作不会从上游同步到下游，进而导致上下游数据不一致。这与旧版本 DM 的行为一致，因为在旧版本中 TiDB 不支持外键。
```

> - v6.6.0 版本的 [DM](/dm/dm-overview.md) 在同步数据到下游 TiDB 时，会显式关闭下游 TiDB 的 `foreign_key_checks`，所以由外键产生级联操作不会从上游同步到下游，进而导致上下游数据不一致。这与旧版本 DM 的行为一致，因为在旧版本中 TiDB 不支持外键。

Reviewer 3: ```suggestion
- [DM](/dm/dm-overview.md) v6.6.0 在同步数据到下游 TiDB 时，会显式关闭下游 TiDB 的 [`foreign_key_checks`](/system-variables.md#foreign_key_checks)，所以由外键产生的级联操作不会从上游同步到下游，进而导致上下游数据不一致。这与 v6.6.0 之前版本 DM 的行为一致，因为在 v6.6.0 之前 TiDB 不支持外键。
```

- v6.6.0 版本的 [TiCDC](/ticdc/ticdc-overview.md) 同步兼容了外键功能。旧版本的 [TiCDC](/ticdc/ticdc-overview.md) 在同步外键的表时，可能会报错，建议在下游 TiDB 关闭 `foreign_key_checks`。

> - v6.6.0 版本的 [BR](/br/backup-and-restore-overview.md) 同步兼容了外键功能。旧版本的 [BR](/br/backup-and-restore-overview.md) 在 restore 外键的表时，可能会报错，建议在下游 TiDB 关闭 `foreign_key_checks` 后再 restore。

Reviewer 5: ```suggestion
- v6.6.0 版本的 [BR](/br/backup-and-restore-overview.md) 同步兼容了外键功能。旧版本的 [BR](/br/backup-and-restore-overview.md) 在 restore 带外键的表到 v6.6.0 以及之后的集群时，可能会报错，建议在下游 TiDB 关闭 `foreign_key_checks` 后再 restore。
```

> - v6.6.0 版本的 [BR](/br/backup-and-restore-overview.md) 同步兼容了外键功能。旧版本的 [BR](/br/backup-and-restore-overview.md) 在 restore 带外键的表到 v6.6.0 以及之后的集群时，可能会报错，建议在下游 TiDB 关闭 `foreign_key_checks` 后再 restore。

Reviewer 3: ```suggestion
- v6.6.0 版本的 [BR](/br/backup-and-restore-overview.md) 兼容外键功能。之前版本的 BR 在恢复带外键的表到 v6.6.0 以及之后的集群时，可能会报错。建议在下游 TiDB 关闭 `foreign_key_checks` 后再恢复集群。
```

> - [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md) 在导入数据到 TiDB 前，建议先关闭 TiDB 的 `foreign_key_checks`。

Reviewer 3: ```suggestion
- 使用 [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md) 导入数据到 TiDB 前，建议先关闭 TiDB 系统变量 `foreign_key_checks`。
```

> - [Dumpling](/dumpling-overview.md) 与外键没有兼容性问题。

Author: nice catch, done.

> - [sync-diff-inspector](/sync-diff-inspector/sync-diff-inspector-overview.md) 在对比上下游数据时，如果上下游数据库的版本不一样，且下游 TiDB 中存在[不生效的外键](#tidb-版本间兼容性)，则 [sync-diff-inspector](/sync-diff-inspector/sync-diff-inspector-overview.md) 可能会报上下游表结构不一致，因为在 v6.6.0 版本之后的 TiDB 会给不生效的外键加一条 `/* FOREIGN KEY INVALID */` 注释。

Reviewer 3: ```suggestion
- [sync-diff-inspector](/sync-diff-inspector/sync-diff-inspector-overview.md) 在对比上下游数据时，如果上下游数据库的版本不一样，且下游 TiDB 中存在[不生效的外键](#tidb-版本间兼容性)，则 sync-diff-inspector 可能会报上下游表结构不一致的错误。因为 TiDB v6.6.0 会给不生效的外键加一条 `/* FOREIGN KEY INVALID */` 注释。
```

### 与 MySQL 的兼容性

> - 创建外键未指定名称时，TiDB 自动生成的外键名称和 MySQL 不一样。例如 TiDB 生成的外键名称为 `fk_1`、`fk_2`、`fk_3` 等，MySQL 生成的外键名称为 `table_name_ibfk_1`、 `table_name_ibfk_2`、`table_name_ibfk_3` 等。

Reviewer 3: ```suggestion
创建外键未指定名称时，TiDB 自动生成的外键名称和 MySQL 不一样。例如 TiDB 生成的外键名称为 `fk_1`、`fk_2`、`fk_3` 等，MySQL 生成的外键名称为 `table_name_ibfk_1`、 `table_name_ibfk_2`、`table_name_ibfk_3` 等。
```

> - [sync-diff-inspector](/sync-diff-inspector/sync-diff-inspector-overview.md) 在对比上下游数据时，如果上下游数据库的版本不一样，且下游 TiDB 中存在[不生效的外键](#tidb-版本间兼容性)，则 sync-diff-inspector 可能会报上下游表结构不一致的错误。因为 TiDB v6.6.0 会给不生效的外键加一条 `/* FOREIGN KEY INVALID */` 注释。

Author: nice catch, done.
